### ISSUE TITLE
Mosh Detached session message

### ISSUE THREAD

**Issue by sn00pster** [SynoCommunity/spksrc#3582]
mosh-server has the ability to display a message telling you of any detached sessions that are connected to the server.

This support is based on utmp and libutempter.  The current build of the most server does not link against libutempter, therefore the functionality is not available.

I have done some very hacky exploratory work, basically fudging into the build a Makefile for utempter which requires a lot of interaction while building to get it to use and link against utempter.

However, the message is still not displayed, so there is something else at fault here.  I will try to discover what it is, but I know nothing about the correct way of getting utempter to automatically build and use a configure script, so I cannot share my work on this as it requires me to manually edit files during building and after failures in order to get the library built.

Any pointers on how I can get libutempter to download and configure/build properly would be helpful and would certainly advance figuring out what is going on.

**Comment by m4tt075** [comment]
I'll be happy to help, if you are willing to do all the testing. Just PR your WIP and we'll take it from there...  ;-)

**Comment by sn00pster** [comment]
Let me try and figure something out, I can’t do a PR on the hacky bit as I’m fudging stuff inside the work-x64 folder as it builds....  I can do a PR on the libutempter “module” under cross and the additional dependency that mosh now has on it.

libutempter does not use a configure script, it’s just a Makefile, I used autoconf to generate a configure script and then once it started building I then replaced the generated Makefile with a modified version of the original makefile from the source package so that the libraries get written to the right place so that when mosh builds its configure script picks up that libutempter is available.....told you it was a hack! Lol

I’m an embedded software engineer by trade and my only PC coding these days is limited to Qt, so I never get involved with makefiles which is why i’ve Struggled to this point.

**Comment by m4tt075** [comment]
To get you started with our build system:
- You will need to create a cross-package for `utempter` in `/cross/utempter`. Then, you will have to add that as a dependency in the `cross/mosh/Makefile` , make it compile and put the libraries in the right place.
- Start creating the `cross/utempter` package first
- You can copy the `Makefile` from `/cross/mosh` for instance to get you started
- If there is no configure script, you can override configuration by adding `CONFIGURE_TARGET = nop" in the `Makefile` and the step will be skipped. I don't think `autoconf` will be needed.
- Pathes can be a bit tricky. Is there an option to define an `INSTALL_PREFIX`?

P.S.: I'm working in marketing these days. If I can do it, you can...  :-)

**Comment by sn00pster** [comment]
Thanks! I spent ages looking for how to disable configure!!

I’ve done cross/utempter
Added the dependency in cross/mosh/Makefile

Will doublecheck the Makefile for installpath.

Thanks, this has been very helpful. I can see a way through way this with these hints!

**Comment by m4tt075** [comment]
Right. OK, you'll have to patch them out. I'd add a patch replacing `usr` by something like an `%install_prefix%` placeholder. That patch will be applied automatically during the build process. Then, you should be able to add some `post_patch_target` to the `Makefile` replacing that placeholder by the `$(STAGING_INSTALL_PREFIX)` variable. This is, where all the dependencies are staged and where `mosh` should be looking. You can refer to `/cross/tvheadend` far an examplatory `post_patch_target` implementation.

Eventually, you will also need some other stuff, i.e. `PLIST` and `digests` files. You'll find the background here: https://github.com/SynoCommunity/spksrc/wiki/Developers-HOW-TO

**Comment by sn00pster** [comment]
Perfect, I’ve been awake for 40 odd hours now so will take this up again tomorrow.  I’ve already done the PLIST and digest file, the patch should be easy as I’ve seen examples in the cross folder and they’re minor modifications.

Thank you very much for your help.

**Comment by m4tt075** [comment]
If you `cd` into the overarching `spk/working-<arch>-<fwvers>` directory and run `ls -latr`, you'll see cookie files, like `.copy.done`, for all the individual build steps and recursions into cross packages, that have been completed. You can remove the cookie files of the build steps you want to repeat.

**Comment by sn00pster** [comment]
Awesome @m4tt075 

Using the information you've provided here, I know have the cross/libutempter "module" fully working patches, build etc.

Should I put together a PR here for the WIP, although it's not currently working, mosh is seeing the library during configuration and building the binary with the appropriate support.

Couldn't have done this without your excellent help!

One step closer.

By the way, I couldn't get the "clean" to work, I deleted the cookies and also the libutempter build folder, but when I did a make again it just didn't do anything and just told me nothing to be done, any further suggestions?

**Comment by m4tt075** [comment]
Glad to hear it went so well and happy to help. You are advancing this project and we appreciate anybody willing to contribute! 

On your question: Yes, it might not work if you cut out intermediate build steps. To make it work, usually you have to remove all the build steps and sub-directories created by a cross package. It might also be necessary to remove all subsequent build steps if there are interdependencies. You'd certainly have to remove those steps consolidating and creating the final package itself as well. I admit it can be tricky...

**Comment by dulitz** [comment]
Thanks a million y'all. Mosh with detached sessions would be completely amazing.

**Comment by sn00pster** [comment]
OK, so appointments finished early today and I decided to torture myself by debugging this on my iPad over a mosh connection to my server.

I now have libutempter working. It requires a few more patches and I will try to clean up the makefile as well as there's a load of stuff in there which was commented on above that isnt actually needed. 

Upshot is that mosh connections now show any detached sessions!  

Will try to sort this out tomorrow when I'm back at a proper keyboard, this was hard work on a touchscreen and sftping files about!

**Comment by sn00pster** [comment]
@ymartin59 good call, I think it makes sense to be a separate package anyway, it's such a great tool. Been very handy today when ssh would have kicked me out quite a number of times.

Glad to have contributed something as minor as the detached session reporting.

**Comment by iFredOS** [comment]
How is it possible to download the separated Mosh package? I'm looking for Mosh for DSM for a few days and found this discussion.

**Comment by chickenandpork** [comment]
I think it may show on https://synocommunity.com/packages when available but for now you might need to get it by installing synocli ( https://synocommunity.com/package/synocli-net ) .. probably better to put the package repos url into your NAS and pull packages through the standard interface though (look for “Easy Install” on https://synocommunity.com/ )

**Comment by sn00pster** [comment]
Bear in mind this conversation, it doesn’t currently build against libutempter so it’s posdible to create a lot of detached sessions and not realise.

**Comment by iFredOS** [comment]
>Bear in mind this conversation, it doesn’t currently build against libutempter so it’s posdible to create a lot of detached sessions and not realise.

@sn00pster OK.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
